lib/tasks/advisory.rake

```diff
@@ -0,0 +1,54 @@
+namespace :advisory do
+  public def remove_unaffected(unaffected_versions)
+    version_array = []
+    unaffected_versions.each do |versions|
+      version_array << if versions.split.length > 2
+                         versions.split(',')
+                       else
+                         versions
+                       end
+    end
+    version_array.flatten!
+    version_array.each do |versions|
+      unaffectedv_array = versions.split
+      unaffected_version_number = unaffectedv_array.second
+      case unaffectedv_array.first
+      when ">="
+        delete_if { |v| Gem::Version.new(v) >= Gem::Version.new(unaffected_version_number) }
+      when "<="
+        delete_if { |v| Gem::Version.new(v) <= Gem::Version.new(unaffected_version_number) }
+      when "~>"
+        delete_if { |v| Gem::Version.new(v) == Gem::Version.new(unaffected_version_number) }
+      when ">"
+        delete_if { |v| Gem::Version.new(v) > Gem::Version.new(unaffected_version_number) }
+      when "<"
+        delete_if { |v| Gem::Version.new(v) < Gem::Version.new(unaffected_version_number) }
+      end
+    end
+  end
+
+  desc 'Mark versions as vulnerable'
+  task update: :environment do
+    begin
+      system "git submodule update"
+      vuln = {}
+      count = 0
+      Dir.chdir(Rails.root.join('storage', 'ruby-advisory-db'))
+      ActiveRecord::Base.transaction do
+        Dir.glob("**/*.yml") do |gem_name|
+          vuln = YAML.load_file(gem_name)
+          rubygem = Rubygem.find_by(name: vuln['gem'])
+          unless rubygem.nil?
+            all_versions = rubygem.versions
+            version_number_list = all_versions.pluck(:number)
+            version_number_list.remove_unaffected((vuln['patched_versions'].to_a + vuln['unaffected_versions'].to_a).flatten)
+            count += all_versions.where(number: version_number_list, vulnerable: false).update_all(vulnerable: true)
+          end
+        end
+      end
```

Comment:
Can you move this code into a class somewhere so it can be tested? Preferably the rake task is just calling one method.

Reply:
Should I move the whole code inside a model `Advisory` and use something like `Advisory.update` inside the rake task to run it? We can then cover the code by writing unit tests for the advisory model.

Reply:
You can create a module in [lib/tasks/helpers](https://github.com/rubygems/rubygems.org/tree/master/lib/tasks/helpers).